influxdb3_catalog/src/catalog.rs

```diff
@@ -743,19 +614,97 @@ impl DatabaseSchema {
-    updated_or_new_tables: &mut SerdeVecMap<TableId, Arc<TableDefinition>>,
-) {
-    if let Some(deleted_table_defn) = deleted_table_defn {
-        if let Some(deleted_table) = updated_or_new_tables.get_mut(&deleted_table_defn.table_id) {
-            let deletion_time = Time::from_timestamp_nanos(deleted_table_defn.deletion_time);
-            let table_name =
-                make_new_name_using_deleted_time(&deleted_table_defn.table_name, deletion_time);
+trait UpdateDatabaseSchema {
+    fn update_schema<'a>(&self, schema: Cow<'a, DatabaseSchema>)
+        -> Result<Cow<'a, DatabaseSchema>>;
+}
+
+impl UpdateDatabaseSchema for CatalogOp {
+    fn update_schema<'a>(
+        &self,
+        schema: Cow<'a, DatabaseSchema>,
+    ) -> Result<Cow<'a, DatabaseSchema>> {
+        match &self {
+            CatalogOp::CreateDatabase(_) => Ok(schema),
+            CatalogOp::CreateTable(create_table) => create_table.update_schema(schema),
+            CatalogOp::AddFields(field_additions) => field_additions.update_schema(schema),
+            CatalogOp::CreateMetaCache(meta_cache_definition) => {
+                meta_cache_definition.update_schema(schema)
+            }
+            CatalogOp::DeleteMetaCache(delete_meta_cache) => {
+                delete_meta_cache.update_schema(schema)
+            }
+            CatalogOp::CreateLastCache(create_last_cache) => {
+                create_last_cache.update_schema(schema)
+            }
+            CatalogOp::DeleteLastCache(delete_last_cache) => {
+                delete_last_cache.update_schema(schema)
+            }
+            CatalogOp::DeleteDatabase(delete_database) => delete_database.update_schema(schema),
+            CatalogOp::DeleteTable(delete_table) => delete_table.update_schema(schema),
+        }
+    }
+}
+
+impl UpdateDatabaseSchema for influxdb3_wal::TableDefinition {
+    fn update_schema<'a>(
+        &self,
+        mut database_schema: Cow<'a, DatabaseSchema>,
+    ) -> Result<Cow<'a, DatabaseSchema>> {
+        match database_schema.tables.get(&self.table_id) {
+            Some(existing_table) => {
+                if let Cow::Owned(updated_table) = existing_table.check_and_add_new_fields(&self)? {
+                    database_schema
+                        .to_mut()
+                        .insert_table(self.table_id, Arc::new(updated_table));
+                }
+            }
+            None => {
+                let new_table = TableDefinition::new_from_op(&self);
+                database_schema
+                    .to_mut()
+                    .insert_table(new_table.table_id, Arc::new(new_table));
+            }
+        }
+        Ok(database_schema)
+    }
+}
+
+impl UpdateDatabaseSchema for DeleteDatabaseDefinition {
+    fn update_schema<'a>(
+        &self,
+        mut schema: Cow<'a, DatabaseSchema>,
+    ) -> Result<Cow<'a, DatabaseSchema>> {
+        // TODO: check if we want to re-delete an already deleted DB. That is current behavior.
```

Comment:
yeah, when we delete a DB, its name gets changed. This way you can immediately create a new db with the same name and it will pass uniqueness validation. If they then issue another delete, they're deleting the new DB, not the old one.